src/FastConv/Convolver.cpp

```diff
@@ -0,0 +1,323 @@
+
+    // do convolution for each block on frequency domain
+    for (int i = 0; i < m_IRNumBlock; ++i) {
+        memcpy(m_X, m_X_origin, sizeof(CFft::complex_t) * doubleBlockLength);
+        __complexVectorMul_I(m_X, m_H_real[i], m_H_imag[i]); // Y[i] = X[i] * H[i], but &Y == &X
+        m_pCFft->doInvFft(iFFTTemp, m_X);
+        CVectorFloat::mulC_I(iFFTTemp, static_cast<float>(doubleBlockLength), doubleBlockLength);
+
+        int offset = i * m_blockLength;
+        for (int j = 0; j < 2; ++j) {
+            __addToRingBuffer(m_buffer->getHead(offset + j * m_blockLength), iFFTTemp + j * m_blockLength, m_blockLength);
+        }
+    }
+
+    // update output using ringbuffer buffer
+    __flushRingBufferToOutput(output, bufferLength);
+
+    applyWetGain(output, bufferLength);
+}
+
+Error_t UniformlyPartitionedFFTConvolver::flushBuffer(float *pfOutputBuffer) {
+    for (int i = 0; i < m_IRNumBlock; ++i) {
+        memcpy(
+            pfOutputBuffer + i * m_blockLength, 
+            m_buffer->getHead(i * m_blockLength), 
+            sizeof(float) * std::min(m_blockLength, m_originIRLengthInSample - 1 - i * m_blockLength)
+        );
+    }
+    applyWetGain(pfOutputBuffer, m_originIRLengthInSample - 1);
+    return Error_t::kNoError;
+}
+
+void UniformlyPartitionedFFTConvolver::__complexVectorMul_I(CFft::complex_t* A, const float* bReal, const float* bImag) {
+    m_pCFft->splitRealImag(aReal, aImag, A);
+
+    __complexVectorMul_I(aReal, aImag, bReal, bImag);
+
+    m_pCFft->mergeRealImag(A, cReal, cImag);
+}
+
+void UniformlyPartitionedFFTConvolver::__complexVectorMul_I(const float* aReal, const float* aImag, const float *bReal, const float* bImag) {
+    // check https://mathworld.wolfram.com/ComplexMultiplication.html for how to reduce 1 multiplication
+
+    // c_r = a_r * b_r - a_i * b_i
+    CVectorFloat::copy (aReal_bReal, aReal, m_blockLengthPlusOne); 
+    CVectorFloat::mul_I(aReal_bReal, bReal, m_blockLengthPlusOne);
+    CVectorFloat::copy (aImag_bImag, aImag, m_blockLengthPlusOne);
+    CVectorFloat::mul_I(aImag_bImag, bImag, m_blockLengthPlusOne);
+    CVectorFloat::copy (cReal, aReal_bReal, m_blockLengthPlusOne);
+    CVectorFloat::sub_I(cReal, aImag_bImag, m_blockLengthPlusOne);
+    
+    // c_i = (a_r + a_i) * (b_r + b_i) - a_r * b_r - a_i * b_i
+    CVectorFloat::copy (cImag, aReal, m_blockLengthPlusOne);
+    CVectorFloat::add_I(cImag, aImag, m_blockLengthPlusOne);
+    CVectorFloat::copy (temp,  bReal, m_blockLengthPlusOne);
+    CVectorFloat::add_I(temp,  bImag, m_blockLengthPlusOne);
+    CVectorFloat::mul_I(cImag, temp,  m_blockLengthPlusOne);
+    CVectorFloat::sub_I(cImag, aReal_bReal, m_blockLengthPlusOne);
+    CVectorFloat::sub_I(cImag, aImag_bImag, m_blockLengthPlusOne);
+}
+
+void UniformlyPartitionedFFTConvolver::__addToRingBuffer(float* bufferHead, float* data, int length) {
+    auto headIndex = bufferHead - m_buffer->begin();
+    if (headIndex + length >= m_buffer->getLength()) {
```

Comment:
Tricky. I wonder if this should be a method on `CRingBuffer`.

Reply:
I have never thought of this! But I think it's weird for ring buffer to provide such method... as the key functionality is to serve as a conceptually circular array.

I think instead of `CRingBuffer`, maybe a separate class like `CircularArray` that focus more on range addition & read & write would be better. At least the code will be way much easier to read.

Reply:
And their conceptual difference should be more easily grasped.